pkg/broker/broker.go

```diff
@@ -15,17 +21,67 @@ var _ brokerapi.ServiceBroker = Broker{}
 // an API server.
 type Broker struct {
 	logger *zap.SugaredLogger
-	atlas  atlas.Client
 }
 
-// NewBroker creates a new Broker with the specified Atlas client and logger.
-func NewBroker(client atlas.Client, logger *zap.SugaredLogger) *Broker {
+// NewBroker creates a new Broker with a logger.
+func NewBroker(logger *zap.SugaredLogger) *Broker {
 	return &Broker{
 		logger: logger,
-		atlas:  client,
 	}
 }
 
+// ContextKey represents the key for a value saved in a context. Linter
+// requires keys to have their own type.
+type ContextKey string
+
+// ContextKeyAtlasClient is the key used to store the Atlas client in the
+// request context.
+var ContextKeyAtlasClient = ContextKey("atlas-client")
+
+// AuthMiddleware is used to validate and parse Atlas API credentials passed
+// using basic auth. The credentials parsed into an Atlas client which is
+// attached to the request context. This client can later be retrieved by the
+// broker from the context.
+func AuthMiddleware(baseURL string) mux.MiddlewareFunc {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			username, password, ok := r.BasicAuth()
+
+			// The username contains both the group ID and public key
+			// formatted as "<PUBLIC_KEY>@<GROUP_ID>".
+			splitUsername := strings.Split(username, "@")
+
+			// If the credentials are invalid we respond with 401 Unauthorized.
+			// The username needs have the correct format and the password must
+			// not be empty.
+			validUsername := len(splitUsername) == 2
+			validPassword := password != ""
+			if !(ok && validUsername && validPassword) {
+				w.WriteHeader(http.StatusUnauthorized)
+				return
+			}
+
+			// Create a new client with the extracted API credentials and
+			// attach it to the request context.
+			client := atlas.NewClient(baseURL, splitUsername[1], splitUsername[0], password)
+			ctx := context.WithValue(r.Context(), ContextKeyAtlasClient, client)
+
+			next.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+// atlasClientFromContext will retrieve an Atlas client stored inside the
+// provided context.
+func atlasClientFromContext(ctx context.Context) (atlas.Client, error) {
```

Comment:
I'm not really convinced that this function provides much value over just inlining its contents where needed. Not big deal though.

Reply:
True! I mostly like the way it reads as part of the individual functions. I feel like `ctx.Value(ContextKeyAtlasClient).(atlas.Client)` is a bit hard on the eyes without context.